### PR TITLE
Bluetooth: controller: Use random numbers in adv and enc setup

### PR DESCRIPTION
**Pull request by cvinayak**
Added implementation to get random numbers in ISR. And fixed
implementation to use random numbers in advertisement random
delay and encryption setup procedure.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

### CODE REVIEW

File: subsys/bluetooth/controller/hal/nrf5/rand.c [hunk 1]
```diff
@@ -12,36 +12,60 @@
 #include "common/log.h"
 #include "hal/debug.h"
 
-#define RAND_RESERVED (4)
+enum rand_status {
```
Comment (carlescufi) [inline subsys/bluetooth/controller/hal/nrf5/rand.c]: Use standard errors here, no need to introduce a new set of errors. Also return -ERR

File: subsys/bluetooth/controller/hal/nrf5/rand.c [hunk 2]
```diff
@@ -12,36 +12,60 @@
...
 	u8_t last;
 	u8_t rand[1];
 };
 
-static struct rand *rng;
+static struct rand *rng_isr;
+static struct rand *rng_thr;
 
-void rand_init(u8_t *context, u8_t context_len)
+static void _init(struct rand **rng, u8_t *context, u8_t len, u8_t threshold)
```
Comment (carlescufi) [inline subsys/bluetooth/controller/hal/nrf5/rand.c]: _ underscore is reserved for the kernel, let's avoid it from now on even if it's static

Comment (carlescufi) [inline subsys/bluetooth/controller/hal/nrf5/rand.c]: Perhaps do_init() ?

File: subsys/bluetooth/controller/hal/nrf5/rand.c [hunk 3]
```diff
@@ -12,36 +12,60 @@
...
+{
+	_init(&rng_thr, context, context_len, threshold);
+}
+
+void rand_isr_init(u8_t *context, u8_t context_len, u8_t threshold)
+{
+	_init(&rng_isr, context, context_len, threshold);
+}
+
+static size_t _get(struct rand *rng, size_t octets, u8_t *rand)
```
Comment (carlescufi) [inline subsys/bluetooth/controller/hal/nrf5/rand.c]: See my comment about underscore prefix

File: subsys/bluetooth/controller/hal/nrf5/rand.c [hunk 4]
```diff
@@ -57,59 +81,85 @@ size_t rand_get(size_t octets, u8_t *rand)
...
+size_t rand_isr_get(size_t octets, u8_t *rand)
+{
+	return _get(rng_isr, octets, rand);
+}
 
-		last = rng->last + 1;
-		if (last == rng->count) {
-			last = 0;
-		}
+static enum rand_status _isr(struct rand *rng, bool store)
```
Comment (carlescufi) [inline subsys/bluetooth/controller/hal/nrf5/rand.c]: See my comment about underscore prefix

File: subsys/bluetooth/controller/ll_sw/ll.c [hunk 5]
```diff
@@ -41,7 +41,10 @@
 /* Global singletons */
 
 /* memory for storing Random number */
-static u8_t MALIGN(4) _rand_context[3 + 4 + 1];
+#define RAND_THREAD_THRESHOLD 4  /* atleast access address */
+#define RAND_ISR_THRESHOLD    12 /* atleast encryption div. and iv */
+static u8_t MALIGN(4) _rand_context[4 + RAND_THREAD_THRESHOLD + 1];
```
Comment (carlescufi) [inline subsys/bluetooth/controller/ll_sw/ll.c]: See my comment about underscore prefix

File: subsys/bluetooth/controller/ll_sw/ll.c [hunk 6]
```diff
@@ -41,7 +41,10 @@
...
+static u8_t MALIGN(4) _rand_isr_context[4 + RAND_ISR_THRESHOLD + 1];
```
Comment (carlescufi) [inline subsys/bluetooth/controller/ll_sw/ll.c]: See my comment about underscore prefix

**Review by carlescufi — CHANGES_REQUESTED** [review]

**Comment by cvinayak** [comment]
@carlescufi updated with review rework

**Review by carlescufi — APPROVED** [review]

**Review by jhedberg — APPROVED** [review]

**Pull request merged**